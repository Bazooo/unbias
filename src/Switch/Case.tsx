import { SwitchElementProps, SwitchPrimitive } from './Switch'

interface SwitchCaseProps extends SwitchElementProps {
  equals: SwitchPrimitive
}

const Case = ({ children }: SwitchCaseProps) => children

Case.switchElementType = 'CASE'

export default Case
