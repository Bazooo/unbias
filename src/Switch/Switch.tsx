import React from 'react'

export type SwitchPrimitive = string | number

export interface SwitchElementProps {
  equals?: SwitchPrimitive
  children: JSX.Element
}

interface SwitchElementType {
  ({ children }: SwitchElementProps): JSX.Element
  switchElementType: 'CASE' | 'DEFAULT_CASE'
}

type SwitchElement = React.ReactElement<SwitchElementProps, SwitchElementType>

interface SwitchProps {
  statement: SwitchPrimitive | (() => SwitchPrimitive)
  children: SwitchElement | SwitchElement[]
}

const Switch = ({ children, statement }: SwitchProps) => {
  const childArray = Array.isArray(children) ? children : [children]

  let caseChild: JSX.Element | null = null
  let defaultChild: JSX.Element | null = null

  childArray.forEach((child, i) => {
    if (caseChild) {
      return
    }

    switch (child.type.switchElementType) {
      case 'CASE':
        if (child.props.equals === statement) {
          caseChild = child.props.children
        }
        break
      case 'DEFAULT_CASE':
        defaultChild = defaultChild || child.props.children
        break
      default:
        throw new Error(`Switch child[${i}] is invalid!`)
    }
  })

  return <>{caseChild || defaultChild}</>
}

export default Switch
