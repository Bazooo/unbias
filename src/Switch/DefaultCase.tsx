import { SwitchElementProps } from './Switch'
import { Omit } from 'type-zoo/types'

type SwitchDefaultCaseProps = Omit<SwitchElementProps, 'equals'>

const DefaultCase = ({ children }: SwitchDefaultCaseProps) => children

DefaultCase.switchElementType = 'DEFAULT_CASE'

export default DefaultCase
