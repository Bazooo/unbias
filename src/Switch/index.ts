import Switch from './Switch'
import Case from './Case'
import DefaultCase from './DefaultCase'

export default Switch
export {
  Case,
  DefaultCase
}
