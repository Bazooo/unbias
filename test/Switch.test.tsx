import React from 'react'
import { shallow } from 'enzyme'

import Switch, { Case, DefaultCase } from '../src/Switch'

describe('Switch', () => {
  it('should show first case on right statement', () => {
    const wrapper = shallow(
      <Switch statement='statement'>
        <Case equals='statement'><span id='right'>test</span></Case>
        <Case equals='statement'><span id='wrong'>test</span></Case>
        <DefaultCase><span id='very-wrong'>test</span></DefaultCase>
      </Switch>
    )

    expect(wrapper.find('span#right').exists()).toBeTruthy()
    expect(wrapper).toMatchSnapshot()
  })

  it('should show first right case', () => {
    const wrapper = shallow(
      <Switch statement='statement'>
        <Case equals='wrong statement'><span id='wrong'>test</span></Case>
        <Case equals='still wrong statement'><span id='still-wrong'>test</span></Case>
        <Case equals='statement'><span id='right'>test</span></Case>
        <Case equals='statement'><span id='kinda-wrong'>test</span></Case>
        <DefaultCase><span id='very-wrong'>test</span></DefaultCase>
      </Switch>
    )

    expect(wrapper.find('span#right').exists()).toBeTruthy()
    expect(wrapper).toMatchSnapshot()
  })

  it('should show first default case', () => {
    const wrapper = shallow(
      <Switch statement='nothing'>
        <Case equals='wrong statement'><span id='wrong'>test</span></Case>
        <Case equals='still wrong statement'><span id='still-wrong'>test</span></Case>
        <Case equals='statement'><span id='wrong3'>test</span></Case>
        <Case equals='statement'><span id='wrong4'>test</span></Case>
        <DefaultCase><span id='right'>test</span></DefaultCase>
        <DefaultCase><span id='kinda-wrong'>test</span></DefaultCase>
      </Switch>
    )

    expect(wrapper.find('span#right').exists()).toBeTruthy()
    expect(wrapper).toMatchSnapshot()
  })

  it('should allow numbers', () => {
    const wrapper = shallow(
      <Switch statement={1}>
        <Case equals={1}><span id='right'>test</span></Case>
        <Case equals={2}><span id='wrong'>test</span></Case>
        <DefaultCase><span id='very-wrong'>test</span></DefaultCase>
      </Switch>
    )

    expect(wrapper.find('span#right').exists()).toBeTruthy()
    expect(wrapper).toMatchSnapshot()
  })

  it('should show null if no default case', () => {
    const wrapper = shallow(
      <Switch statement='statement'>
        <Case equals='wrong-statement'><span id='wrong'>test</span></Case>
      </Switch>
    )

    expect(wrapper.find('span#wrong').exists()).toBeFalsy()
    expect(wrapper).toMatchSnapshot()
  })

  it('should not accept wrong element in the switch', () => {
    const wrapper = () => shallow(
      <Switch statement='nothing'>
        <span>wrong</span>
      </Switch>
    )

    expect(wrapper).toThrowError('Switch child[0] is invalid!')
  })

  it('works', () => {
    const wrapper = shallow(
      <Switch statement='statement'>
        <Case equals='statement'><span id='first'>test</span></Case>
        <Case equals='statement'><span id='second'>test</span></Case>
        <Case equals={1}><span id='third'>test</span></Case>
        <DefaultCase><span id='default'>test</span></DefaultCase>
      </Switch>
    )

    expect(wrapper).toMatchSnapshot()

    expect(wrapper.find('span#first').exists()).toBeTruthy()
    expect(wrapper.find('span#second').exists()).toBeFalsy()
    expect(wrapper.find('span#third').exists()).toBeFalsy()
    expect(wrapper.find('span#default').exists()).toBeFalsy()

    wrapper.setProps({ statement: 1 })

    expect(wrapper.find('span#first').exists()).toBeFalsy()
    expect(wrapper.find('span#second').exists()).toBeFalsy()
    expect(wrapper.find('span#third').exists()).toBeTruthy()
    expect(wrapper.find('span#default').exists()).toBeFalsy()

    wrapper.setProps({ statement: 'none' })

    expect(wrapper.find('span#first').exists()).toBeFalsy()
    expect(wrapper.find('span#second').exists()).toBeFalsy()
    expect(wrapper.find('span#third').exists()).toBeFalsy()
    expect(wrapper.find('span#default').exists()).toBeTruthy()
  })
})
