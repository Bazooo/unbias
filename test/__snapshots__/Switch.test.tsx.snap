// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch should allow numbers 1`] = `
<Fragment>
  <span
    id="right"
  >
    test
  </span>
</Fragment>
`;

exports[`Switch should show first case on right statement 1`] = `
<Fragment>
  <span
    id="right"
  >
    test
  </span>
</Fragment>
`;

exports[`Switch should show first default case 1`] = `
<Fragment>
  <span
    id="right"
  >
    test
  </span>
</Fragment>
`;

exports[`Switch should show first right case 1`] = `
<Fragment>
  <span
    id="right"
  >
    test
  </span>
</Fragment>
`;

exports[`Switch should show null if no default case 1`] = `<Fragment />`;

exports[`Switch works 1`] = `
<Fragment>
  <span
    id="first"
  >
    test
  </span>
</Fragment>
`;
